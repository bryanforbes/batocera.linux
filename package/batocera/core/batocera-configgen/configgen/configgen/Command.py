from __future__ import annotations

from typing import TYPE_CHECKING

if TYPE_CHECKING:
    from collections.abc import Mapping, Sequence
    from pathlib import Path


class Command:
    def __init__(self, array: Sequence[str | Path], env: Mapping[str, str | Path] = {}):
        self.array = list(array)
        self.env = dict(env)

    def __str__(self):
        strings = [f"{varName}={varValue}" for varName, varValue in self.env.items()]

        strings.extend(str(value) for value in self.array)

        return " ".join(strings)
