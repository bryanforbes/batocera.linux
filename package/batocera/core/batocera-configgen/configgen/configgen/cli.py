from __future__ import annotations

import logging
import time
from collections.abc import Callable
from functools import partial
from pathlib import Path
from typing import Any, TypedDict, TypeVar, Unpack, cast

import click

from . import profiler
from .emulators import run_emulator
from .utils.logger import setup_logging

_FC = TypeVar('_FC', bound=Callable[..., Any] | click.Command)

_logger = logging.getLogger(__name__)


def _player_callback(player: int, ctx: click.Context, option: click.Option, value: int | None) -> int | None:
    current_value: list[PlayerOptions] = ctx.params.setdefault('players', [])

    option_names = (
        f'p{player}hat_count',
        f'p{player}button_count',
        f'p{player}device_path',
        f'p{player}real_name',
        f'p{player}guid',
        f'p{player}index',
    )

    items = [
        (f'p{player}axis_count', value),
        *[(option_name, ctx.params.pop(option_name)) for option_name in option_names],
    ]

    if all(item[1] is not None for item in items):
        p_length = len(f'p{player}')
        current_value.append(
            cast(
                PlayerOptions,
                {
                    'player_number': player,
                    **{option_name[p_length:]: option_value for option_name, option_value in items},
                },
            )
        )
        return value

    if all(item[1] is None for item in items):
        return value

    raise click.UsageError(f'All options that start with p{player} must be specified if any are specified', ctx=ctx)


def _players(max_players: int) -> Callable[[_FC], _FC]:
    def wrap(callable: _FC) -> _FC:
        for n in range(max_players, 0, -1):
            callable = click.option(f'-p{n}nbaxes', f'p{n}axis_count', type=int, callback=partial(_player_callback, n))(
                callable
            )
            callable = click.option(f'-p{n}nbhats', f'p{n}hat_count', type=int)(callable)
            callable = click.option(f'-p{n}nbbuttons', f'p{n}button_count', type=int)(callable)
            callable = click.option(f'-p{n}devicepath', f'p{n}device_path', type=str)(callable)
            callable = click.option(f'-p{n}name', f'p{n}real_name', type=str)(callable)
            callable = click.option(f'-p{n}guid', type=str)(callable)
            callable = click.option(f'-p{n}index', type=int)(callable)

        return callable

    return wrap


class PlayerOptions(TypedDict):
    player_number: int
    index: int
    guid: str
    real_name: str
    device_path: str
    button_count: int
    hat_count: int
    axis_count: int


class Options(TypedDict):
    system: str
    rom: Path
    emulator: str | None
    core: str | None
    netplaymode: str | None
    netplaypass: str | None
    netplayip: str | None
    netplayport: str | None
    netplaysession: str | None
    state_slot: str | None
    state_filename: Path | None
    autosave: str | None
    systemname: str | None
    gameinfoxml: Path
    lightgun: bool
    wheel: bool
    trackball: bool
    spinner: bool
    players: list[PlayerOptions]


@click.command()
@click.option('-system', required=True)
@click.option('-rom', required=True, type=click.Path(path_type=Path))
@click.option('-emulator')
@click.option('-core')
@click.option('-netplaymode')
@click.option('-netplaypass')
@click.option('-netplayip')
@click.option('-netplayport')
@click.option('-netplaysession')
@click.option('-state_slot')
@click.option('-state_filename', type=click.Path(path_type=Path))
@click.option('-autosave')
@click.option('-systemname')
@click.option('-gameinfoxml', default='/dev/null', type=click.Path(path_type=Path))
@click.option('-lightgun', is_flag=True)
@click.option('-wheel', is_flag=True)
@click.option('-trackball', is_flag=True)
@click.option('-spinner', is_flag=True)
@_players(8)
def cli(**kwargs: Unpack[Options]) -> None:
    with setup_logging():
        exit_code = -1

        try:
            exit_code = run_emulator(kwargs)
        except Exception:
            _logger.exception('configgen exception: ')

        profiler.disable(dump=True)

        time.sleep(1)  # this seems to be required so that the gpu memory is restituated and available for es
        _logger.debug('Exiting configgen with status %s', exit_code)

        exit(exit_code)
