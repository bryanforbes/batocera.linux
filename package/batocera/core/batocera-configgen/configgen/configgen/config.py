from __future__ import annotations

import logging
import xml.etree.ElementTree as ET
from collections.abc import Iterator, Mapping
from dataclasses import dataclass
from typing import TYPE_CHECKING, Any, ClassVar, Literal, Self, TypeVar, overload

import yaml

from .batoceraPaths import BATOCERA_SHADERS, DEFAULTS_DIR, ES_SETTINGS, USER_SHADERS
from .utils.missing import MISSING, MissingType

if TYPE_CHECKING:
    from pathlib import Path

    from .cli import Options
    from .settings.unixSettings import UnixSettings

_T = TypeVar('_T')
_TV = TypeVar('_TV')
_FV = TypeVar('_FV')

_logger = logging.getLogger(__name__)


# adapted from https://gist.github.com/angstwad/bf22d1822c38a92ec0a9
def _dict_merge(destination: dict[str, Any], source: Mapping[str, Any]) -> None:
    """Recursive dict merge. Inspired by :meth:``dict.update()``, instead of
    updating only top-level keys, dict_merge recurses down into dicts nested
    to an arbitrary depth, updating keys. The ``merge_dct`` is merged into
    ``dct``.
    :param destination: dict onto which the merge is executed
    :param source: dict merged into destination
    :return: None
    """
    for key, value in source.items():
        if key in destination and isinstance(destination[key], dict) and isinstance(value, Mapping):
            _dict_merge(destination[key], value)
        else:
            destination[key] = value


def _load_defaults(system: str, default_file: Path, default_arch_file: Path, /) -> dict[str, Any]:
    with default_file.open() as f:
        defaults: dict[str, Any] = yaml.load(f, Loader=yaml.CLoader)

    with default_arch_file.open() as f:
        arch_defaults: dict[str, Any] = yaml.load(f, Loader=yaml.CLoader) or {}

    config: dict[str, Any] = defaults.get('default', {})

    if 'default' in arch_defaults:
        _dict_merge(config, arch_defaults['default'])

    if system in defaults:
        _dict_merge(config, defaults[system])

    if system in arch_defaults:
        _dict_merge(config, arch_defaults[system])

    return config


@dataclass(slots=True)
class Config:
    TRUE_VALUES: ClassVar[set[Literal['1', 'true', 'on', 'enabled', True]]] = {'1', 'true', 'on', 'enabled', True}
    MISSING: ClassVar = MISSING

    data: dict[str, Any]

    def __contains__(self, x: object, /) -> bool:
        return x in self.data

    @overload
    def get(self, key: str, default: MissingType = ..., /) -> Any | MissingType: ...

    @overload
    def get(self, key: str, default: _T, /) -> Any | _T: ...

    def get(self, key: str, default: _T | MissingType = MISSING, /) -> Any | _T | MissingType:
        return self.data.get(key, default)

    @overload
    def get_bool(self, key: str, default: bool = False, /, *, return_values: None = None) -> bool: ...

    @overload
    def get_bool(self, key: str, default: bool = False, /, *, return_values: tuple[_TV, _FV]) -> _TV | _FV: ...

    def get_bool(
        self, key: str, default: bool = False, /, *, return_values: tuple[_TV, _FV] | None = None
    ) -> bool | _TV | _FV:
        value = self.get(key)

        if value is MISSING:
            return default

        if isinstance(value, str):
            value = value.lower()

        if return_values is None:
            return value in self.TRUE_VALUES

        return return_values[value not in self.TRUE_VALUES]

    @overload
    def get_str(self, key: str, default: MissingType = ..., /) -> str | MissingType: ...

    @overload
    def get_str(self, key: str, default: str, /) -> str: ...

    def get_str(self, key: str, default: str | MissingType = MISSING, /) -> str | MissingType:
        value = self.get(key)

        if value is MISSING:
            return default

        return str(value)

    @overload
    def get_int(self, key: str, default: MissingType = ..., /) -> int | MissingType: ...

    @overload
    def get_int(self, key: str, default: int, /) -> int: ...

    def get_int(self, key: str, default: int | MissingType = MISSING, /) -> int | MissingType:
        value = self.get(key)

        if value is MISSING:
            return default

        return int(value)

    @overload
    def get_float(self, key: str, default: MissingType = ..., /) -> float | MissingType: ...

    @overload
    def get_float(self, key: str, default: float, /) -> float: ...

    def get_float(self, key: str, default: float | MissingType = MISSING, /) -> float | MissingType:
        value = self.get(key)

        if value is MISSING:
            return default

        return float(value)

    def items(self, /, *, starts_with: str | None = None) -> Iterator[tuple[str, Any]]:
        if starts_with is None:
            yield from self.data.items()
        else:
            starts_with_len = len(starts_with)
            for key, value in self.data.items():
                if key.startswith(starts_with):
                    yield key[starts_with_len:], value


def _get_rom_settings_name(rom: Path, /) -> str:
    # sanitize rule by EmulationStation
    # see FileData::getConfigurationName() on batocera-emulationstation
    rom_settings_name = rom.name.replace('=', '').replace('#', '')

    _logger.info('game settings name: %s', rom_settings_name)

    return rom_settings_name


@dataclass
class SystemConfig(Config):
    @property
    def core(self) -> str:
        return self.data['core']

    @property
    def core_forced(self) -> str:
        return self.data['core-forced']

    @property
    def emulator(self) -> str:
        return self.data['emulator']

    @property
    def emulator_forced(self) -> str:
        return self.data['emulator-forced']

    @property
    def ui_mode(self) -> Literal['Full', 'Kiosk', 'Kid']:
        return self.data['uimode']

    @property
    def show_fps(self) -> bool:
        return self.data['showFPS']

    @property
    def video_mode(self) -> str:
        return self.data.get('videomode', 'default')

    @property
    def use_guns(self) -> bool:
        return self.get_bool('use_guns')

    @property
    def use_wheels(self) -> bool:
        return self.get_bool('use_wheels')

    @classmethod
    def load(cls, options: Options, settings: UnixSettings, /) -> Self:
        defaults = _load_defaults(
            options['system'], DEFAULTS_DIR / 'configgen-defaults.yml', DEFAULTS_DIR / 'configgen-defaults-arch.yml'
        )
        data: dict[str, Any] = {
            'emulator': defaults['emulator'],
            'core': defaults['core'],
        }

        # In the yaml files, the "options" structure is not flat, so we have to flatten it here
        # because the options are flat in batocera.conf to make it easier for end users to edit
        if 'options' in defaults:
            _dict_merge(data, defaults['options'])

        rom = options['rom']
        rom_settings_name = _get_rom_settings_name(rom)

        data.update(settings.get_all_iter('display', keep_name=True, keep_defaults=True))
        data.update(settings.get_all_iter('controllers', keep_name=True))

        global_settings = settings.get_all('global')
        system_settings = settings.get_all(options['system'])
        folder_settings = settings.get_all(f'{options["system"]}.folder["{rom.parent}"]')
        game_settings = settings.get_all(f'{options["system"]}["{rom_settings_name}"]')

        data.update(global_settings)
        data.update(system_settings)
        data.update(folder_settings)
        data.update(game_settings)

        try:
            es_config = ET.parse(ES_SETTINGS)

            # showFPS
            drawframerate_node = es_config.find("./bool[@name='DrawFramerate']")
            drawframerate_value = drawframerate_node.attrib['value'] if drawframerate_node is not None else 'false'
            if drawframerate_value not in ['false', 'true']:
                drawframerate_value = 'false'

            data['showFPS'] = drawframerate_value == 'true'

            # uimode
            uimode_node = es_config.find("./string[@name='UIMode']")
            uimode_value = uimode_node.attrib['value'] if uimode_node is not None else 'Full'
            if uimode_value not in ['Full', 'Kiosk', 'Kid']:
                uimode_value = 'Full'

            data['uimode'] = uimode_value
        except Exception:
            data['showFPS'] = False
            data['uimode'] = 'Full'

        data['emulator-forced'] = (
            'emulator' in global_settings
            or 'emulator' in system_settings
            or 'emulator' in game_settings
            or options['emulator'] is not None
        )
        data['core-forced'] = (
            'core' in global_settings
            or 'core' in system_settings
            or 'core' in game_settings
            or options['core'] is not None
        )

        if options['emulator'] is not None:
            data['emulator'] = options['emulator']

        if options['core'] is not None:
            data['core'] = options['core']

        if 'use_guns' not in data and options['lightgun']:
            data['use_guns'] = True

        if 'use_wheels' not in data and options['wheel']:
            data['use_wheels'] = True

        # network options
        if options['netplaymode'] is not None:
            data['netplay.mode'] = options['netplaymode']

        if options['netplaypass'] is not None:
            data['netplay.password'] = options['netplaypass']

        if options['netplayip'] is not None:
            data['netplay.server.ip'] = options['netplayip']

        if options['netplayport'] is not None:
            data['netplay.server.port'] = options['netplayport']

        if options['netplaysession'] is not None:
            data['netplay.server.session'] = options['netplaysession']

        # autosave arguments
        if options['state_slot'] is not None:
            data['state_slot'] = options['state_slot']

        if options['autosave'] is not None:
            data['autosave'] = options['autosave']

        if options['state_filename'] is not None:
            data['state_filename'] = options['state_filename']

        return cls(data)


@dataclass
class RenderConfig(Config):
    @classmethod
    def load(cls, options: Options, settings: UnixSettings, shader_set: str | MissingType, /) -> Self:
        data: dict[str, Any] = {}

        if shader_set is not cls.MISSING:
            if shader_set == 'none':
                rendering_defaults = BATOCERA_SHADERS / 'configs' / 'rendering-defaults.yml'
            else:
                rendering_defaults = USER_SHADERS / 'configs' / shader_set / 'rendering-defaults.yml'
                if not rendering_defaults.exists():
                    rendering_defaults = BATOCERA_SHADERS / 'configs' / shader_set / 'rendering-defaults.yml'

            data = _load_defaults(
                options['system'], rendering_defaults, rendering_defaults.with_name('rendering-defaults-arch.yml')
            )

        rom = options['rom']
        rom_settings_name = _get_rom_settings_name(rom)

        # es only allow to update system and game settings in fact for the moment

        # for compatibility with earlier Batocera versions, let's keep -renderer
        # but it should be reviewed when we refactor configgen (to Python3?)
        # so that we can fetch them from system.shader without -renderer
        data.update(settings.get_all_iter(f'{options["system"]}-renderer'))
        data.update(settings.get_all_iter(f'{options["system"]}["{rom_settings_name}"]-renderer'))

        return cls(data)
