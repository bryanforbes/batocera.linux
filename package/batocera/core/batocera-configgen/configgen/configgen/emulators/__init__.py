from __future__ import annotations

from contextlib import nullcontext
from importlib import import_module
from typing import TYPE_CHECKING, Final

from ..batoceraPaths import BATOCERA_CONF
from ..config import RenderConfig, SystemConfig
from ..controller import Controller
from ..controllersConfig import getGamesMetaData
from ..gun import Gun, precalibrate_guns
from ..settings.unixSettings import UnixSettings
from ..utils.squashfs import squashfs_rom
from ..utils.wheels import configure_wheels
from ._base import Emulator as Emulator  # noqa: TCH001

if TYPE_CHECKING:
    from ..cli import Options

_GENERATOR_MAP: Final[dict[str, str]] = {}


def get_emulator_cls(emulator: str, /) -> type[Emulator]:
    module_path = _GENERATOR_MAP.get(emulator, emulator.replace('-', '_'))

    try:
        module = import_module(f'..{module_path}', package=__name__)
        emulator_cls: type[Emulator] = module.Emulator
    except ImportError as e:
        if e.name is not None and e.name.startswith(__name__.split('.')[0]):
            raise Exception(f'no emulator found for {emulator!r}') from e

        raise
    except AttributeError as e:
        raise Exception(f'no emulator found for {emulator!r}') from e

    return emulator_cls


def run_emulator(options: Options, /) -> int:
    exit_code = -1
    original_rom = options['rom']

    if original_rom.suffix == '.squashfs':
        rom_context_manager = squashfs_rom(original_rom)
    else:
        rom_context_manager = nullcontext(original_rom)

    with rom_context_manager as rom:
        system = options['system']

        settings = UnixSettings(BATOCERA_CONF)
        system_config = SystemConfig.load(options, settings)
        render_config = RenderConfig.load(options, settings, system_config.get('shaderset'))

        metadata = getGamesMetaData(system, original_rom)

        controllers = Controller.load_for_player_options(options['players'])

        if system_config.use_guns:
            guns = Gun.get_all()
            precalibrate_guns(system, system_config.emulator, system_config.core, original_rom)
        else:
            guns = {}

        with configure_wheels(controllers, system, system_config, metadata) as (controllers, wheels):
            emulator_cls = get_emulator_cls(system_config.emulator)
            emulator = emulator_cls(system, system_config, render_config, metadata, controllers, guns, wheels)
            exit_code = emulator.run(rom, original_rom, options['gameinfoxml'])

    return exit_code
