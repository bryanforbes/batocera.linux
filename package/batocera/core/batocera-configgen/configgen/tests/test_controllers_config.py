from __future__ import annotations

from pathlib import Path
from typing import TYPE_CHECKING

import pytest

from configgen.batoceraPaths import ES_GAMES_METADATA
from configgen.controllersConfig import (
    getAssociatedMouse,
    getDevicesInformation,
    getGamesMetaData,
    shortNameFromPath,
)

if TYPE_CHECKING:
    from unittest.mock import Mock

    from pyfakefs.fake_filesystem import FakeFilesystem
    from pytest_mock import MockerFixture
    from syrupy.assertion import SnapshotAssertion

    from configgen.types import DeviceInfoMapping


@pytest.mark.parametrize(
    'path',
    [
        '/path/to/ABCDEFGHIGKLMNOPQRSTUVWXYZ!@#$%^&*abcdefghigklmnopqrstuvwxyz!@#$%^&*0123456789!@#$%^&*.zip',
        Path('/another/path/to/AbCdEf@#%!0&.bar.rom'),
        Path('/another/path/to/rom-name [US] suffix.rom'),
        Path('/another/path/to/rom-name (something) suffix.rom'),
        '/userdata/roms/system/blah (something [US]) [EN] (spam) foo.dsk',
    ],
    ids=lambda x: f'Path({x!s})' if isinstance(x, Path) else x,
)
def test_short_name_from_path(path: str | Path, snapshot: SnapshotAssertion) -> None:
    assert shortNameFromPath(path) == snapshot


@pytest.mark.parametrize(
    'rom',
    [
        '/path/to/wheel rom [US].zip',
        Path('/path/to/trackball#rom (US).zip'),
    ],
    ids=lambda x: f'Path({x!s})' if isinstance(x, Path) else x,
)
@pytest.mark.parametrize(
    'system',
    [
        'naomi',
        'naomi2',
        'atomiswave',
        'fbneo',
        'mame',
        'neogeo',
        'triforce',
        'hypseus-singe',
        'model2',
        'model3',
        'hikaru',
        'gaelco',
        'cave3rd',
        'namco2x6',
        'dreamcast',
        'snes',
    ],
)
def test_get_games_meta_data(fs: FakeFilesystem, system: str, rom: str | Path, snapshot: SnapshotAssertion) -> None:
    fs.create_file(
        ES_GAMES_METADATA,
        contents="""<?xml version="1.0" encoding="UTF-8"?>
<systems xmlns:xsi="http://www.w3.org/2001/XMLSchema-instance" xsi:noNamespaceSchemaLocation="gamesdb.xsd">
  <system id="arcade">
    <game id="wheelrom">
      <wheel one="1"/>
    </game>
    <game id="trackballrom">
      <trackball/>
    </game>
  </system>
  <system id="dreamcast">
    <game id="wheelrom">
      <wheel two="2"/>
    </game>
  </system>
  <system id="snes">
    <game id="default">
      <wheel three="3" four="4"/>
    </game>
    <game id="wheelrom">
      <wheel four="42" five="5"/>
    </game>
    <game id="trackballrom">
      <gun foo="bar"/>
    </game>
    <game />
  </system>
  <system />
</systems>
""",
    )

    assert getGamesMetaData(system, rom) == snapshot


def test_get_games_meta_data_no_systems(fs: FakeFilesystem) -> None:
    fs.create_file(
        ES_GAMES_METADATA,
        contents="""<?xml version="1.0" encoding="UTF-8"?>
<systems xmlns:xsi="http://www.w3.org/2001/XMLSchema-instance" xsi:noNamespaceSchemaLocation="gamesdb.xsd">
</systems>
""",
    )

    assert getGamesMetaData('dreamcast', 'rom one.zip') == {}


@pytest.mark.parametrize(
    'rom',
    [
        'rom one.zip',
        'romTwo.zip',
    ],
)
@pytest.mark.parametrize(
    'system',
    [
        'naomi',
        'naomi2',
        'atomiswave',
        'fbneo',
        'mame',
        'neogeo',
        'triforce',
        'hypseus-singe',
        'model2',
        'model3',
        'hikaru',
        'gaelco',
        'cave3rd',
        'namco2x6',
        'dreamcast',
        'snes',
    ],
)
def test_get_games_meta_data_no_metadata(fs: FakeFilesystem, system: str, rom: str) -> None:
    fs.create_file(
        ES_GAMES_METADATA,
        contents="""<?xml version="1.0" encoding="UTF-8"?>
<systems xmlns:xsi="http://www.w3.org/2001/XMLSchema-instance" xsi:noNamespaceSchemaLocation="gamesdb.xsd">
  <system id="arcade">
    <game id="romone">
    </game>
  </system>
  <system id="dreamcast">
    <game id="default"/>
  </system>
</systems>
""",
    )

    assert getGamesMetaData(system, rom) == {}


def test_get_devices_information(fs: FakeFilesystem, mocker: MockerFixture, pyudev: Mock) -> None:
    fs.create_file('/sys/devices/event1/actual_device')
    fs.create_symlink('/sys/devices/event1/device/device', link_target='/sys/devices/event1/actual_device')
    fs.create_file('/sys/devices/event2/actual_device')
    fs.create_file('/sys/devices/event3/actual_device')
    fs.create_file('/sys/devices/event5/actual_device')
    fs.create_file('/sys/devices/event6/actual_device')

    context_instance = mocker.Mock()

    context_instance.list_devices.return_value = [
        mocker.Mock(device_node='/dev/foo'),
        mocker.Mock(device_node='/dev/input/event0', properties={}),
        mocker.Mock(
            device_node='/dev/input/event2',
            properties={'ID_INPUT_TOUCHPAD': '1'},
            sys_path='/sys/devices/event2',
        ),
        mocker.Mock(device_node='/dev/input/event4', properties={'ID_INPUT_WHEEL': '1'}),
        mocker.Mock(
            device_node='/dev/input/event3',
            properties={'ID_INPUT_JOYSTICK': '1', 'ID_PATH': 'foo-bar'},
            sys_path='/sys/devices/event3',
        ),
        mocker.Mock(
            device_node='/dev/input/event1', properties={'ID_INPUT_MOUSE': '1'}, sys_path='/sys/devices/event1'
        ),
        mocker.Mock(
            device_node='/dev/input/event5',
            properties={'ID_INPUT_JOYSTICK': '1', 'ID_INPUT_WHEEL': '1', 'ID_PATH': 'foo-bar'},
            sys_path='/sys/devices/event5',
        ),
        mocker.Mock(
            device_node='/dev/input/event6',
            properties={
                'ID_INPUT_JOYSTICK': '1',
                'ID_INPUT_WHEEL': '1',
                'ID_PATH': 'foo-bar-baz',
                'WHEEL_ROTATION_ANGLE': '4',
            },
            sys_path='/sys/devices/event6',
        ),
    ]

    pyudev.Context.return_value = context_instance

    assert getDevicesInformation() == {
        '/dev/input/event1': {
            'eventId': 1,
            'sysfs_path': '/sys/devices/event1/actual_device',
            'isJoystick': False,
            'isWheel': False,
            'isMouse': True,
            'associatedDevices': None,
            'joystick_index': None,
            'mouse_index': 0,
        },
        '/dev/input/event2': {
            'eventId': 2,
            'sysfs_path': '/sys/devices/event2/device/device',
            'isJoystick': False,
            'isWheel': False,
            'isMouse': True,
            'associatedDevices': None,
            'joystick_index': None,
            'mouse_index': 1,
        },
        '/dev/input/event3': {
            'eventId': 3,
            'sysfs_path': '/sys/devices/event3/device/device',
            'isJoystick': True,
            'isWheel': False,
            'isMouse': False,
            'associatedDevices': ['/dev/input/event5'],
            'joystick_index': 0,
            'mouse_index': None,
        },
        '/dev/input/event5': {
            'eventId': 5,
            'sysfs_path': '/sys/devices/event5/device/device',
            'isJoystick': True,
            'isWheel': True,
            'isMouse': False,
            'associatedDevices': ['/dev/input/event3'],
            'joystick_index': 1,
            'mouse_index': None,
        },
        '/dev/input/event6': {
            'eventId': 6,
            'sysfs_path': '/sys/devices/event6/device/device',
            'isJoystick': True,
            'isWheel': True,
            'isMouse': False,
            'associatedDevices': [],
            'joystick_index': 2,
            'mouse_index': None,
            'wheel_rotation': 4,
        },
    }


def test_get_devices_information_no_devices(mocker: MockerFixture, pyudev: Mock) -> None:
    context_instance = mocker.Mock()

    context_instance.list_devices.return_value = [
        mocker.Mock(device_node='/dev/foo'),
        mocker.Mock(device_node='/dev/input/event0', properties={}),
        mocker.Mock(device_node='/dev/input/event4', properties={'ID_INPUT_WHEEL': '1'}),
    ]

    pyudev.Context.return_value = context_instance

    assert getDevicesInformation() == {}


@pytest.mark.parametrize(
    ('mapping', 'expected'),
    [
        pytest.param({}, None, id='empty mapping'),
        pytest.param({'/dev/input/event0': {}}, None, id='device not in mapping'),
        pytest.param({'/dev/input/event3': {'associatedDevices': None}}, None, id='associatedDevices is None'),
        pytest.param({'/dev/input/event3': {'associatedDevices': []}}, None, id='associatedDevices is empty'),
        pytest.param(
            {
                '/dev/input/event3': {'associatedDevices': ['/dev/input/event1']},
                '/dev/input/event1': {'isMouse': False},
            },
            None,
            id='associated device is not a mouse',
        ),
        pytest.param(
            {
                '/dev/input/event3': {'associatedDevices': ['/dev/input/event1', '/dev/input/event2']},
                '/dev/input/event1': {'isMouse': False},
                '/dev/input/event2': {'isMouse': True},
            },
            '/dev/input/event2',
            id='associated device is a mouse',
        ),
    ],
)
def test_get_associated_mouse(mapping: DeviceInfoMapping, expected: str | None) -> None:
    assert getAssociatedMouse(mapping, '/dev/input/event3') == expected
