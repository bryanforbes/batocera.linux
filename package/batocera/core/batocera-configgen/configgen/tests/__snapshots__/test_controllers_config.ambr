# serializer version: 1
# name: test_get_games_meta_data[atomiswave-/path/to/wheel rom [US].zip]
  dict({
    'wheel_one': '1',
  })
# ---
# name: test_get_games_meta_data[atomiswave-Path(/path/to/trackball#rom (US).zip)]
  dict({
  })
# ---
# name: test_get_games_meta_data[cave3rd-/path/to/wheel rom [US].zip]
  dict({
    'wheel_one': '1',
  })
# ---
# name: test_get_games_meta_data[cave3rd-Path(/path/to/trackball#rom (US).zip)]
  dict({
  })
# ---
# name: test_get_games_meta_data[dreamcast-/path/to/wheel rom [US].zip]
  dict({
    'wheel_two': '2',
  })
# ---
# name: test_get_games_meta_data[dreamcast-Path(/path/to/trackball#rom (US).zip)]
  dict({
  })
# ---
# name: test_get_games_meta_data[fbneo-/path/to/wheel rom [US].zip]
  dict({
    'wheel_one': '1',
  })
# ---
# name: test_get_games_meta_data[fbneo-Path(/path/to/trackball#rom (US).zip)]
  dict({
  })
# ---
# name: test_get_games_meta_data[gaelco-/path/to/wheel rom [US].zip]
  dict({
    'wheel_one': '1',
  })
# ---
# name: test_get_games_meta_data[gaelco-Path(/path/to/trackball#rom (US).zip)]
  dict({
  })
# ---
# name: test_get_games_meta_data[hikaru-/path/to/wheel rom [US].zip]
  dict({
    'wheel_one': '1',
  })
# ---
# name: test_get_games_meta_data[hikaru-Path(/path/to/trackball#rom (US).zip)]
  dict({
  })
# ---
# name: test_get_games_meta_data[hypseus-singe-/path/to/wheel rom [US].zip]
  dict({
    'wheel_one': '1',
  })
# ---
# name: test_get_games_meta_data[hypseus-singe-Path(/path/to/trackball#rom (US).zip)]
  dict({
  })
# ---
# name: test_get_games_meta_data[mame-/path/to/wheel rom [US].zip]
  dict({
    'wheel_one': '1',
  })
# ---
# name: test_get_games_meta_data[mame-Path(/path/to/trackball#rom (US).zip)]
  dict({
  })
# ---
# name: test_get_games_meta_data[model2-/path/to/wheel rom [US].zip]
  dict({
    'wheel_one': '1',
  })
# ---
# name: test_get_games_meta_data[model2-Path(/path/to/trackball#rom (US).zip)]
  dict({
  })
# ---
# name: test_get_games_meta_data[model3-/path/to/wheel rom [US].zip]
  dict({
    'wheel_one': '1',
  })
# ---
# name: test_get_games_meta_data[model3-Path(/path/to/trackball#rom (US).zip)]
  dict({
  })
# ---
# name: test_get_games_meta_data[namco2x6-/path/to/wheel rom [US].zip]
  dict({
    'wheel_one': '1',
  })
# ---
# name: test_get_games_meta_data[namco2x6-Path(/path/to/trackball#rom (US).zip)]
  dict({
  })
# ---
# name: test_get_games_meta_data[naomi-/path/to/wheel rom [US].zip]
  dict({
    'wheel_one': '1',
  })
# ---
# name: test_get_games_meta_data[naomi-Path(/path/to/trackball#rom (US).zip)]
  dict({
  })
# ---
# name: test_get_games_meta_data[naomi2-/path/to/wheel rom [US].zip]
  dict({
    'wheel_one': '1',
  })
# ---
# name: test_get_games_meta_data[naomi2-Path(/path/to/trackball#rom (US).zip)]
  dict({
  })
# ---
# name: test_get_games_meta_data[neogeo-/path/to/wheel rom [US].zip]
  dict({
    'wheel_one': '1',
  })
# ---
# name: test_get_games_meta_data[neogeo-Path(/path/to/trackball#rom (US).zip)]
  dict({
  })
# ---
# name: test_get_games_meta_data[snes-/path/to/wheel rom [US].zip]
  dict({
    'wheel_five': '5',
    'wheel_four': '42',
    'wheel_three': '3',
  })
# ---
# name: test_get_games_meta_data[snes-Path(/path/to/trackball#rom (US).zip)]
  dict({
    'gun_foo': 'bar',
    'wheel_four': '4',
    'wheel_three': '3',
  })
# ---
# name: test_get_games_meta_data[triforce-/path/to/wheel rom [US].zip]
  dict({
    'wheel_one': '1',
  })
# ---
# name: test_get_games_meta_data[triforce-Path(/path/to/trackball#rom (US).zip)]
  dict({
  })
# ---
# name: test_short_name_from_path[/path/to/ABCDEFGHIGKLMNOPQRSTUVWXYZ!@#$%^&*abcdefghigklmnopqrstuvwxyz!@#$%^&*0123456789!@#$%^&*.zip]
  'abcdefghigklmnopqrstuvwxyzabcdefghigklmnopqrstuvwxyz0123456789'
# ---
# name: test_short_name_from_path[/userdata/roms/system/blah (something [US]) [EN] (spam) foo.dsk]
  'blahfoo'
# ---
# name: test_short_name_from_path[Path(/another/path/to/AbCdEf@#%!0&.bar.rom)]
  'abcdef0bar'
# ---
# name: test_short_name_from_path[Path(/another/path/to/rom-name (something) suffix.rom)]
  'romnamesuffix'
# ---
# name: test_short_name_from_path[Path(/another/path/to/rom-name [US] suffix.rom)]
  'romnamesuffix'
# ---
